File: elixir/lib/symphony/workflow.ex
defmodule Symphony.Workflow do
  @moduledoc """
  Parses WORKFLOW.md files: YAML front matter between `---` markers + Markdown body (Liquid template).
  """

  defstruct [:config, :template, :source_path]

  @type t :: %__MODULE__{
          config: map(),
          template: String.t(),
          source_path: String.t() | nil
        }

  @doc "Parse a WORKFLOW.md file from disk."
  def parse_file(path) do
    case File.read(path) do
      {:ok, content} ->
        case parse(content) do
          {:ok, workflow} -> {:ok, %{workflow | source_path: path}}
          error -> error
        end

      {:error, reason} ->
        {:error, {:file_error, reason, path}}
    end
  end

  @doc "Parse WORKFLOW.md content string."
  def parse(content) when is_binary(content) do
    case split_front_matter(content) do
      {:ok, yaml_str, template} ->
        case YamlElixir.read_from_string(yaml_str) do
          {:ok, config} ->
            {:ok, %__MODULE__{config: config, template: String.trim(template)}}

          {:error, reason} ->
            {:error, {:yaml_error, reason}}
        end

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp split_front_matter(content) do
    case String.split(content, ~r/^---\s*$/m, parts: 3) do
      [_before, yaml, body] ->
        {:ok, yaml, body}

      _ ->
        {:error, :no_front_matter}
    end
  end
end

File: src/workflow.ts
import { Effect, Data } from "effect"
import { parse as parseYaml } from "yaml"
import { readFileSync } from "node:fs"

export interface Workflow {
  readonly config: Record<string, unknown>
  readonly template: string
  readonly sourcePath: string | null
}

export class WorkflowParseError extends Data.TaggedError("WorkflowParseError")<{
  readonly reason: string
}> {}

/** Parse a WORKFLOW.md content string into config + template. */
export function parseWorkflow(
  content: string,
  sourcePath: string | null = null,
): Effect.Effect<Workflow, WorkflowParseError> {
  return Effect.try({
    try: () => {
      const parts = content.split(/^---\s*$/m)
      if (parts.length < 3) {
        throw new Error("No YAML front matter found (expected --- delimiters)")
      }
      const yamlStr = parts[1]!
      const template = parts.slice(2).join("---").trim()
      const config = parseYaml(yamlStr) as Record<string, unknown>
      return { config, template, sourcePath }
    },
    catch: (err) =>
      new WorkflowParseError({
        reason: err instanceof Error ? err.message : String(err),
      }),
  })
}

/** Parse a WORKFLOW.md file from disk. */
export function parseWorkflowFile(
  path: string,
): Effect.Effect<Workflow, WorkflowParseError> {
  return Effect.flatMap(
    Effect.try({
      try: () => readFileSync(path, "utf-8"),
      catch: (err) =>
        new WorkflowParseError({
          reason: `Failed to read ${path}: ${err instanceof Error ? err.message : String(err)}`,
        }),
    }),
    (content) => parseWorkflow(content, path),
  )
}
